Add Playwright e2e test suite ready for incremental migration and visual regression testing

Jinksi opened the pull request:
Resolves #8177

#### Changes proposed in this Pull Request

This PR lays the foundation for Playwright e2e and visual regression testing, ready for incremental migration of existing Puppeteer e2e tests.

Migrating our e2e tests to use Playwright instead of Puppeteer will provide us benefits including, but not limited to:
* Reduced test flakiness/increased test stability with [auto-waiting](https://playwright.dev/docs/actionability) and configurable test [retries](https://playwright.dev/docs/test-retries).
* Easier to write robust and maintainable e2e tests with [locators](https://playwright.dev/docs/locators#quick-guide) for targeting DOM elements using testing-library APIs (consistent with integration test APIs).
* [Visual regression testing](https://playwright.dev/docs/test-snapshots) to help us update component library dependencies with more confidence.


To aid in migration, this PR follows the [WooCommerce Puppeteer to Playwright migration guide](https://github.com/woocommerce/woocommerce/wiki/Puppeteer-to-Playwright-Migration-Guide-for-e2e-Testing) which adds a Playwright test suite alongside the existing Puppeteer test suite.

**TODO**

- [x] Migrate example tests
- [x] Setup GH check
- [x] Add basic tips on how to write PlayWright tests in `tests/e2e-pw/README.md`
- [x] #8160 
- [x] Review and merge this PR 🚀
- [x] Discuss handover with Harmony – FE guild will remain caretakers until a future time TBD
- [x] Post the draft P2 post (paJDYF-cjc-p2) to inform stakeholder teams and encourage adoption (link to `tests/e2e-pw/README.md` for getting started)


**Notes/inclusions:**

* The fundamental Playwright boilerplate and config.
* Authentication setup step for admin & customer with caching for local dev envs.
* Some initial basic tests with visual regression snapshots.
* Uses the existing e2e docker environment setup.
* Uses many of the same conventions from the [WC Puppeteer to Playwright migration guide](https://github.com/woocommerce/woocommerce/wiki/Puppeteer-to-Playwright-Migration-Guide-for-e2e-Testing) (e.g. `e2e-pw` directory).
* A GH action that runs e2e-pw on PRs to `develop` or `trunk` and outputs the Playwright HTML report as a GH action artefact ([see this workflow run](https://github.com/Automattic/woocommerce-payments/actions/runs/7218882881?pr=7922))

<img width="1038" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/ce1d4bb1-d69c-4e1a-afc5-5185388873b7">


#### Testing instructions
1. `npm install`
1. Setup e2e environment
	1. `npm run test:e2e-setup`
	1. `npm run test:e2e-up`
1. Run e2e tests with Playwright
	-   `npm run test:e2e-pw` headless run from within a linux docker container.
	-   `npm run test:e2e-pw-ui` runs tests in interactive UI mode from within a linux docker container – recommended for authoring tests and re-running failed tests.
	-   `npm run test:e2e-pw keyword` runs tests only with a specific keyword in the file name, e.g. `dispute` or `checkout`.
	-   `npm run test:e2e-pw --update-snapshots` updates snapshots.
1. Ensure the GH check `E2E Playwright Tests` passes
	- View the summary of the GH check run and download the `playwright-report.zip`
	- Open the `index.html` locally in your browser to view the test report
1. Ensure Puppeteer tests pass

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests
- [x] Tested on mobile **N/A**

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Initialise playwright config

Jinksi committed:
Init visual regression testing

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7671` or branch name `try/playwright-e2e-tests` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=try/playwright-e2e-tests) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9cd9095f5f452f988676fe413251ce9060476533
- Build time: 2024-04-19 01:24:10 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Add merchant deposits spec

Jinksi committed:
Load Payments Overview in basic.spec.ts

Jinksi committed:
Make e2e config a ts file

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

haszari commented:
> This PR is a prototype of using Playwright for e2e and visual regression testing.

@Jinksi can you add a little summary of the rationale, pitch or benefits of the switch (Puppeteer => Playwright) for WooPayments?

Likely some crossover with Woo core's rationale, but I'm keen to understand how this benefits us (e.g. maybe aligning with Woo is part of the appeal!).

> There were a number of reasons for this migration:

> Increase the reliability/decrease the flakiness of the end-to-end test suite
Simplify the code base, eliminate a good deal of boilerplate code in favour of built-in functions/conventions
Speed up the test suite execution
Position the test suite to use some of the advanced features available in Playwright (cross browser/device, visual testing) - [Woo Puppeteer to Playwright migration guide](https://github.com/woocommerce/woocommerce/wiki/Puppeteer-to-Playwright-Migration-Guide-for-e2e-Testing)

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Update authentication setup and use helper functions

Jinksi committed:
Update package-lock.json

Jinksi committed:
Ignore temp auth files

Jinksi committed:
Remove basic test visual snapshot

Jinksi committed:
Update deposits snapshot

Jinksi committed:
Re-use authentication state for local dev

Jinksi committed:
Fix inconsistent auth file paths

Jinksi committed:
Avoid `waitForLoadState( 'networkidle' )`

Jinksi committed:
Remove flaky visual regression snapshot test

ismaeldcom committed:
Restore deposits visual regression snapshot test

Take a partial screenshot focusing only on the WCPay page element to avoid flakiness caused by general notices and banners.

ismaeldcom committed:
Initial shopper-checkout-purchase.spec.ts migration

To review required changes and refactors to also get rid of @woocommerce/e2e-utils

Jinksi commented on tests/e2e-pw/playwright.config.ts:
```diff
@@ -0,0 +1,88 @@
+		toHaveScreenshot: { maxDiffPixels: 100 },
+	},
+	snapshotPathTemplate:
+		'{testDir}/__snapshots__/{testFilePath}/{arg}_{projectName}_{platform}{ext}',
```
`{platform}` results in snapshots that aren't compatible across darwin/linux.

E.g. on macOS, the filename is `Merchant-deposits-Load-the-deposits-list-page-1_chromium_darwin.png`
When the test runs on GH actions, it expects a file `Merchant-deposits-Load-the-deposits-list-page-1_chromium_linux.png`

Should we remove the `{platform}` part of snapshot filenames? This would require screenshots being close enough across both darwin/linux.

Jinksi replied:
I'll test this in 6897f72180a6babfe9ecf584d386dd6e3a9ba239 & 97606bf to see if the e2e-pw GH check on #7922 pass

haszari replied:
Could also recommend/require running tests in known env (e.g. docker container running linux).

Jinksi replied:
It might work, but the GH check e2e env has deposit history that the local env doesn't have:

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/8e537c41-c138-4b17-8deb-764099fb8202)

haszari replied:
Aha, so ideally that history would be standardised in the tests container (sounds like this snapshot depends on leaky state and could be more self-contained).

ismaeldcom replied:
>Should we remove the {platform} part of snapshot filenames? This would require screenshots being close enough across both darwin/linux.

Yep, I think we can get rid of the platform part aiming for a common look. We can always adjust the pixel diff to fit slight differences by fonts or similar things.

>It might work, but the GH check e2e env has deposit history that the local env doesn't have

🤔 That makes it trickier, ideally, we should try to have the same env. In the past I used a fully mocked server for visual regression testing, to remove it from the equation, but that will require another big effort. We can try to limit screenshots to more generic parts, for now.

Jinksi replied:
> We can try to limit screenshots to more generic parts, for now.

Yeh, agree we can keep it simple for the MVP of the Playwright PR.

Jinksi replied:
I've replaced the deposit list visual snapshot test with a non-state-dependant UI test in a25a555 – mainly to see how different this will be cross-platform, but also this UI was an element that broke in the past due to a wp/components update IIRC.

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/19727c17-505a-40e8-8e32-1876964f292d)

Jinksi replied:
> see how different this will be cross-platform

The [snapshot test failed](https://github.com/Automattic/woocommerce-payments/actions/runs/7498234661?pr=7671) due to small font-rendering differences. 

```
3533 pixels (ratio 0.01 of all image pixels) are different.
```

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/e7c581e0-d264-4262-89d2-e1a258b27efc)

Linux:
![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/e8f62434-0415-4f99-8179-3c2805d49bca)

Darwin:
![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/827ce09d-bf0a-421e-a879-f6e65ba06c79)

Jinksi replied:
I've adjusted the `maxDiffPixelRatio` to `0.025` in 889b5b5 (tried `0.0005` in b9cc207 but it failed cross-platform). If it becomes flakey we should run these visual regression tests within a linux container as @haszari mentioned.

<img width="592" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/f9135c0a-c5f3-4c85-a6ff-7a14e8767d0d">

ismaeldcom replied:
Yes, I was expecting fonts to be an issue between platforms. I usually solve that by using a specific font instead of the system one. Do you know if we could use a linux container and still use the Playwright UI? 🤔

Jinksi replied:
[Running in UI mode from within a Docker container is mentioned in the docs](https://playwright.dev/docs/test-ui-mode#docker--github-codespaces), so might be worth a shot if we think it'd be more reliable – hopefully it doesn't add too much complexity.

There is a [docs page with the recommended path for running tests from browsers in a Docker container](https://playwright.dev/docs/docker).

Jinksi replied:
Getting the network right is proving painful with Docker:

Running this command is almost working, and the PW UI is available at `localhost:8077`.

```sh
docker run -v -p 8077:8077 $PWD:/woopayments -w /woopayments -e BASE_URL=http://host.docker.internal:8084 -it --rm --ipc=host mcr.microsoft.com/playwright:v1.39.0-jammy npm run test:e2e-pw -- --ui-host=0.0.0.0 --ui-port=8077
```

But then PW only gets a buggy version of WP and cannot login to wp-admin. I think we need to tell WP that the site's hostname is `host.docker.internal:8084`.


![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/b543f96d-b428-4d97-a481-eb45d7c33b0d)

Jinksi replied:
As mentioned in Slack, I think we leave the Docker env part out of this PR's scope and handle it as a follow-up issue. Everything works without running in Docker, although it may make the visual regression tests more reliable.

p1704929040720799/1702046252.009489-slack-C053VQPD74J

Jinksi replied:
Revisiting visual regression testing, I added one for this dispute notice component – font rendering differences will make visual testing unreliable until we fix this.

Should we fix this for this MVP PR or leave it as a followup issue?

<img width="966" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/14054f3a-f127-4e28-8c4e-3595f59104bf">

Jinksi replied:
🏗️ I've got to a point where running in a Docker env locally works, and the visual snapshots match with the GH Actions snapshot.

To do this:

1. I added the following to `tests/e2e/docker/wordpress/wp-config.php`:

	```php
	define('DOCKER_REQUEST_URL', (!empty($_SERVER['HTTPS']) ? 'https://' : 'http://') . (!empty($_SERVER['HTTP_HOST']) ? $_SERVER['HTTP_HOST'] : 'localhost'));
	define('WP_SITEURL', DOCKER_REQUEST_URL);
	define('WP_HOME', DOCKER_REQUEST_URL);
	```

2. Then Run the following commands:
	- UI mode: `docker run -v $PWD:/woopayments -w /woopayments -p 8077:8077 -e BASE_URL=http://host.docker.internal:8084 -it --rm --ipc=host mcr.microsoft.com/playwright:v1.39.0-jammy npm run test:e2e-pw -- --ui-host=0.0.0.0 --ui-port=8077`
	- CLI run: `docker run -v $PWD:/woopayments -w /woopayments -e BASE_URL=http://host.docker.internal:8084 -it --rm --ipc=host mcr.microsoft.com/playwright:v1.39.0-jammy npm run test:e2e-pw`


So if we tidy this up into the npm setup & run scripts, we'll have a better setup for robust visual regression tests.

Jinksi replied:
This thread is getting long, and the Docker part of this may not make it into this MVP PR.

So I've opened #8160 to implement running PW locally from a linux Docker container.

Jinksi replied:
Resolved now that #8160 has been merged into this branch.

Jinksi committed:
Share snapshots across darwin/linux

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Rename deposit snapshot file

Jinksi committed:
Exploration: Playwright GH action workflow (#7922)

Co-authored-by: Jordan Bordonado <jordan.bordonado@automattic.com>

Jinksi commented:
I've merged the GH action PR #7922 into this branch to simplify the work-in-progress into a single PR.

Jinksi committed:
Remove temp trigger branch used for testing GH checks

Jinksi committed:
Replace state-dependant deposit visual test with limited UI test

Jinksi commented on .github/workflows/e2e-pw-pull-request.yml:
```diff
@@ -0,0 +1,77 @@
+              run: |
+                  echo "::group::Install Playwright"
+                  npx playwright install
+                  echo "::endgroup::"
```
This `npx playwright install` step currently installs more browsers than just Chromium – might be wasteful unless we intend to test against the other browsers.

<img width="1091" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/57c98222-7a24-483c-baf1-2b41ccb6f045">

ismaeldcom replied:
I also noticed it on local. We can be more specific and run it like `npx playwright install chromium` to ensure it only installs a single browser.

Jinksi replied:
Resolved in fd60c3780d461d408b1f7789ed3f16e0002960fe

<img width="1052" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/8430718e-f811-4cda-b832-7a9fda16583c">

Jinksi committed:
Increase pixel ratio diff to 0.025 to allow for cross-platform UI inconsistencies

Jinksi committed:
Reduce pixel ratio diff to 0.0005 in order to catch true UI bugs

Jinksi committed:
Revert "Reduce pixel ratio diff to 0.0005 in order to catch true UI bugs"

This reverts commit b9cc207f8d8058b3187edf07562b8abe1890a9d0.

ismaeldcom committed:
Migrate merchant payment gateways disable confirmation

ismaeldcom committed:
Simplify shopper addCartProduct method

Jinksi committed:
Remove redundant comments from playwright config

Jinksi committed:
Only install chromium browser in PW GH action

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Add changelog entry

Jinksi commented on tests/e2e-pw/utils/shopper.ts:
```diff
@@ -0,0 +1,131 @@
+
+	// 3DS 1 cards have another iframe enclosing the authorize form
+	if ( cardType.toUpperCase() === '3DS' ) {
+		const acsFrameHandle = await challengeFrame.waitForSelector(
+			'iframe[name="acsFrame"]'
+		);
```
@ismaeldcom  This selector is causing this test to fail. I think it has changed to `name="stripe-challenge-frame"` but yet to verify.

```
Error: frame.waitForSelector: Page closed
=========================== logs ===========================
waiting for locator('iframe[name="acsFrame"]') to be visible
============================================================
```

```suggestion
		const acsFrameHandle = await challengeFrame.waitForSelector(
			'iframe[name="stripe-challenge-frame"]'
		);
```

View test report and trace [here](https://lighthearted-baklava-76fd5a.netlify.app/#?testId=d5355fd83eb21895cd99-11495a16b0f15e6259cc) 

🌟 Note: playwright-report.zip GH check artefact makes this such an easy way to investigate test failures.

Jinksi replied:
Welp, after reading the 3DS test util I realise `stripe-challenge-frame` isn't going to solve the maze of nested iframes.

Jinksi replied:
I've attempted to resolve this in 896c367, which fixes the failing test when run locally.

Feel free to revert my changes if not suitable.

Jinksi replied:
🙈 Nope, looks like more tests are failing now...

Although they are failing due to WooPayments being disabled in `should disable WCPay after confirming, then enable again without confirming` and not re-enabling.

Jinksi replied:
They passed on re-run!

I guess the test `should disable WCPay after confirming, then enable again without confirming` might be a bit flaky.

Jinksi replied:
I notice a related problem. Since PW is running everything in [parallel](https://playwright.dev/docs/test-parallel) when run locally, some merchant tests (e.g. disabling WooPayments) will cause others to fail (e.g. trying to complete checkout while the plugin is disabled).

Due to this, I ~think we need to~ have disabled parallelism locally as well as in CI.

Jinksi commented:
Note @ismaeldcom: I've added todos to this PR description, I think this covers what we need for this FE guild project launch but feel free to edit ✅

Jinksi committed:
Use `github` reporter when running on GitHub Actions CI

Jinksi committed:
Use `github` AND `html` reporter when running on GitHub Actions CI

Jinksi commented:
I've added the `github` reporter to the PW config in a30a227. This adds annotations to the "Files changed" tab which may help us identify the cause of test failures more quickly.

Jinksi committed:
Expand the README with more details and FAQs

Jinksi committed:
Update README command descriptions

Jinksi committed:
Use `frameLocator` for nested 3DS iframe interaction

Jinksi committed:
Update instructions for viewing test reports

Jinksi committed:
Invalidate auth state after 3 hrs to avoid local auth errors

Jinksi committed:
Disable local parallelism

Jinksi committed:
Increase `expect` timeout from 5s → 10s due to flaky local checkout performance

Jinksi committed:
Add merchant utils file

Jinksi committed:
Add `getShopper` and `getMerchant` functions to allow switching between user contexts

Jinksi committed:
Add merchant disputes order notice spec

Jinksi committed:
Update README with tip about switching between `shopper`/`merchant`

Jinksi commented on tests/e2e-pw/README.md:
````diff
@@ -0,0 +1,88 @@
+).toHaveScreenshot();
+```
+
+**How can I act as shopper or merchant in a test?**
+
+1. A test suite (`describe`) can act as `shopper` or `merchant` by using the helper function `useShopper` or `useMerchant` from `tests/e2e-pw/utils/helpers.ts`.
+2. To switch between `shopper` and `merchant` role, use the `getShopper` and `getMerchant` function:
+
+```ts
+import { getShopper, getMerchant } from './utils/helpers';
+
+test( 'should do things as shopper and merchant', async ( { browser } ) => {
+	const { shopperPage } = await getShopper( browser );
+	const { merchantPage } = await getMerchant( browser );
+
+	// do things as shopper
+	await shopperPage.goto( '/cart/' );
+
+	// do things as merchant
+	await merchantPage.goto( '/wp-admin/admin.php?page=wc-settings' );
+} );
+```
````
While migrating the merchant disputed order notice test, I noticed the problem of switching between `shopper` and `merchant` within a test.

I added util functions to access `shopperPage` and `merchantPage`. What do we think about this pattern?

reykjalin replied:
Love this, it can be really tricky to manage the roles in a test and making that straight-forward is essential ❤️

Jinksi committed:
Update README with instructions for running tests based on keyword

Jinksi committed:
Add dispute details notice visual regression test

Jinksi committed:
Fix grammar

Jinksi committed:
Add link to e2e-setup instructions

Jinksi committed:
Update dispute notice visual snapshot to match linux font rendering

Jinksi committed:
Re-enable WooPayments payment gateway if test fails

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Use Docker container for running Playwright e2e tests locally to make visual comparisons more robust (#8160)

Jinksi committed:
Remove GH workflow temp branch

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi commented on tests/e2e-pw/specs/merchant/merchant-payment-gateways-confirmation.spec.ts:
```diff
@@ -0,0 +1,115 @@
+		await getToggle( page ).click();
+		await waitForToggleLoading( page );
+		await getSaveButton( page ).click();
+		await expect( getToggle( page ) ).toHaveText( 'Yes' );
```
@ismaeldcom this test is being reported as flaky by PW, works on retry.

```
  Locator: getByRole('link', { name: '"WooPayments" payment method is currently' })
    Expected string: "Yes"
    Received string: ""
```

ismaeldcom replied:
Hey @Jinksi can you provide a link to the flaky report? I've tried to find it, but seems to be fine in the latest GH runs.

Jinksi replied:
Yeh, you're right – no flakiness on recent runs 👏 😉

Jinksi committed:
Remove unused arg `cardType` to fix JS Lint warning

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

achyuthajoy commented on .github/workflows/e2e-pw-pull-request.yml:
```diff
@@ -0,0 +1,77 @@
+            - name: Install Playwright
+              shell: bash
+              run: |
+                  echo "::group::Install Playwright"
+                  npx playwright install chromium
+                  echo "::endgroup::"
```
@Jinksi Any particular reason to use groups here? Since this step doesn't use composite actions and the step is doing only one thing, I believe we can remove the group.

Jinksi replied:
Thanks, this was probably just a symptom of copy-pasta.

Jinksi replied:
Resolved in b09866a

achyuthajoy commented on tests/e2e-pw/playwright.config.ts:
```diff
@@ -0,0 +1,58 @@
+export default defineConfig( {
+	testDir: './specs/',
+	/* Run tests in files in parallel */
+	fullyParallel: true,
```
Shouldn't this be `false` since we're not running tests in parallel and limiting the number of workers to 1?

Jinksi replied:
💯 resolved in d8940c1

achyuthajoy commented on tests/e2e-pw/test-e2e-pw-ui.sh:
```diff
@@ -0,0 +1,14 @@
+
+# When updating the Playwright version in the image tage below, make sure to update the npm `@playwright/test` package.json version as well.
+
+docker run -v "$PWD:/woopayments" -w /woopayments \
+	-p 8077:8077 \
+	-e BASE_URL=http://host.docker.internal:8084 \
+	-it --rm --ipc=host \
+	mcr.microsoft.com/playwright:v1.39.0-jammy \
+	npx playwright test --config=tests/e2e-pw/playwright.config.ts --ui --ui-host=0.0.0.0 --ui-port=8077 "$@"
```
Do you think it's better to have the docker config outside the shell script? Like the ones we have for Puppeteer E2E tests?

achyuthajoy commented on tests/e2e-pw/test-e2e-pw.sh:
```diff
@@ -0,0 +1,11 @@
+
+# When updating the Playwright version in the image tage below, make sure to update the npm `@playwright/test` package.json version as well.
+
+docker run -v "$PWD:/woopayments" -w /woopayments \
+	-e BASE_URL=http://host.docker.internal:8084 \
+	-it --rm --ipc=host \
+	mcr.microsoft.com/playwright:v1.39.0-jammy \
+	npx playwright test --config=tests/e2e-pw/playwright.config.ts "$@"
```
Do you think it's better to have the docker config outside the shell script? Like the ones we have for Puppeteer E2E tests?

Jinksi replied:
👍 I agree, I've had a crack at this via a `docker-compose.yml` file in 660c077, let me know what you think.

Jinksi replied:
~I assume this has broken the PW GH action – I'll look into it.~

All E2E tests are failing due to #8291.

Jinksi replied:
Now resolved 🙂

achyuthajoy reviewed (changes requested):
Thanks for working on this @Jinksi. I'm yet to fully test the changes, but left a few comments.

Jinksi committed:
Define `fullyParellel: false` to disable parallel test running

Jinksi committed:
Remove redundant group when running Install Playwright CI step

Jinksi committed:
Use docker-compose file for configuring Playwright docker

Jinksi committed:
Upgrade Playwright 1.39.0 → 1.42.0

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Fix sticky header covering deposit filter visual snapshot

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi commented:
E2E failures due to #8317 – I'll update the card details in the e2e-pw config file once changes have been merged to `develop`.

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Update e2e-pw config card expiry dates

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

vbelolapotkov committed:
Increase single test timeout to 60s

vbelolapotkov committed:
Refactor auth as admin test locators

vbelolapotkov committed:
Extract WP admin login sequence into helper function

vbelolapotkov commented:
@Jinksi thanks for all the work here! I've increased default timeout to 60 secs due to one of the tests failing locally. I checked the UI mode and noticed that setup takes 15-18s out of default 30 for the whole test, and 2-3s extra was needed to complete `should navigate to dispute details when disputed order notice button clicked` test. 

The tests passed locally like a charm.

I've also refactored a bit auth tests hope you don't mind. If you don't like refactors, feel free to revert them.

vbelolapotkov committed:
Fix password input locator to avoid locating show password button

vbelolapotkov committed:
Improve selector for dashboard title

Jinksi commented:
> I've also refactored a bit auth tests hope you don't mind. If you don't like refactors, feel free to revert them.

Thanks for taking this for a spin and making these additions @vbelolapotkov 🙌

The auth setup tests was mostly copy-pasta from [WC core](https://github.com/woocommerce/woocommerce/blob/b10be3d8269dfa22a1be21e4da89cc4979553dac/plugins/woocommerce/tests/e2e-pw/global-setup.js#L64-L70) and your changes to using user-facing locators are an improvement.

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

reykjalin commented on tests/e2e-pw/test-e2e-pw-ui.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
> [!NOTE]
> This is a nitpick.

I think it's typically better to set the crunch bang line to `#!/usr/bin/env bash` (i.e. to use `env`) to make sure `bash` is run in your typical shell environment 🤔 

To be fair though, this didn't really affect the testing for me, but it's what we've done for all our other bash scripts, so I think it'd be better to do this here as well, even if only for consistency 👍

Jinksi replied:
Nice catch! Resolved in e808ea2

reykjalin commented on tests/e2e-pw/specs/shopper/shopper-checkout-purchase.spec.ts:
```diff
@@ -0,0 +1,43 @@
+
+test.describe( 'Successful purchase', () => {
+	test.beforeEach( async ( { page } ) => {
+		await page.goto( '/cart/' );
+		await shopper.addCartProduct( page );
```
Shouldn't `shopper.addCartProduct( page )` come first here? I suppose it doesn't _really_ matter, but I think that's a more typical flow? It's a bit odd to go first to the cart page, then to the product page, and then directly to checkout :)

Jinksi replied:
cc @ismaeldcom as the author of shopper tests.

ismaeldcom replied:
Good catch! It's a leftover from a different code approach, we don't need to visit the cart with this one 🙂

_Addressed in 7b88a7c_

reykjalin commented on tests/e2e-pw/test-e2e-pw.sh:
```diff
@@ -0,0 +1,6 @@
+
+echo "🎭 Running Playwright e2e tests in default headless mode.";
+
+docker compose -f ./tests/e2e-pw/docker-compose.yml run --rm -it --service-ports playwright \
+	npx playwright test --config=tests/e2e-pw/playwright.config.ts "$@"
```
> [!NOTE]
> This is a suggestion for future improvement and doesn't necessarily have to be a part of this PR.

One thing I would loooove to see here is a way to run the tests against some arbitrary URLs, if possible. In other words, to remove the dependency on the Docker containers.

As an example, you could run the `npm task` as something like

```sh
npm run test:e2e-pw -- --store_url=http://wcpay.local --server_url=http://wcpayserver.local
```

This enables 3 things:

- No dependency on Docker
- The ability to switch in a different server, e.g. prod, your sandbox, or someone else's exposed environment.
- The ability to switch in a different store, e.g. a production store, or someone else's exposed dev environment.

We have a setup like this for WooPay so we can run the e2e tests there against our local environments and staging without having separate npm tasks.

reykjalin reviewed (commented):
I have some observations, but haven't been able to run the tests yet because `npm run test:e2e-setup` fails at the following stage:

```
$ npm run test:e2e-setup

# ...

===> Configuring server with stripe account
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress. There has been a critical error on this website.
```

I don't think that's related to this PR so it's probably something in my environment but I haven't been able to figure out what it is. Maybe y'all have run into this error before?

Jinksi committed:
Ensure e2e-pw script run in user env consistent with existing scripts

Jinksi commented:
Thanks for taking a look @reykjalin 🙌

> `npm run test:e2e-setup` fails
> I don't think that's related to this PR so it's probably something in my environment 

It shouldn't be impacted by this PR, how does this script go on `develop`?

I'd suggest running `npm run test:e2e-reset` and `npm run test:e2e-setup` if you haven't already.

ismaeldcom committed:
Remove unneeded shopper test line

reykjalin reviewed (commented):
This is all looking really good! Got this working in [Rancher desktop](https://rancherdesktop.io/) too, so you don't even need Docker 🙌 

`npm run test:e2e-pw-ui` only showed the `auth` tests though, not sure why that's happening?

Jinksi commented:
> `npm run test:e2e-pw-ui` only showed the auth tests though, not sure why that's happening?

@reykjalin This _might_ happen when restarting the container (re-running `npm run test:e2e-pw-ui`). A new `ws` key is generated for the UI each time.

E.g. `localhost:8077/trace/uiMode.html?ws=7bebab541dcfed2157412f1461a5bca7`

You can fix this by visited `http://localhost:8077/` without the `ws` key.

reykjalin commented:
> You can fix this by visited http://localhost:8077/ without the `ws` key

Hmm, that doesn't seem to work 🤔 . `npm run test:e2e-pw` still runs all 13 tests just fine, but `npm run test:e2e-pw-ui` only has the auth tests. Not sure why tbh, but might not be super important?

ismaeldcom commented:
Hey @reykjalin, that's probably because [Playwright UI only shows the first project by default](https://github.com/microsoft/playwright/issues/21952#issuecomment-1483349947), which is the setup part in our case.

reykjalin reviewed (changes requested):
> Hey @reykjalin, that's probably because [Playwright UI only shows the first project by default](https://github.com/microsoft/playwright/issues/21952#issuecomment-1483349947), which is the setup part in our case.

Oh, that's exactly it! I didn't even think to click that, and even if I had I wouldn't have thought to click "chromium". Is there a way we can change the name of the project?

<img width="250" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/13835680/27519c4f-5236-4a25-be54-0674306d9e2b">

***

Also, it seems that `basic.spec.ts` and `auth.setup.ts` serve a very similar purpose. Or at least they're basically the same tests. It's probably enough to keep only one of those? Or somehow merge them?

ismaeldcom commented:
> Is there a way we can change the name of the project?

I think we could [rename it here](https://github.com/Automattic/woocommerce-payments/blob/7b88a7c2f283e9ca39f91fc7ef009d44bcf5f437/tests/e2e-pw/playwright.config.ts#L51), AFAIK that's the actual name, the browser comes from the `use` property.

Jinksi committed:
Use merchant/shopper project names

Jinksi commented:
Reading the [docs](https://playwright.dev/docs/api/class-testproject#properties), PW projects lean towards being used to designate testing for multiple browsers, hence the default `chromium` project.

But I've attempted a `merchant`/`shopper`/`setup`/`basic` project outline in 07e31b64351bb79a92ed1466761885e2568d35d0. This results in the snapshots moving into the `/specs/merchant|shopper` dir, which is all good.

<img width="244" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/df1a593a-89b7-4026-b667-4388360d49c7">

>  It's probably enough to keep only one of those? Or somehow merge them?

I think `basic` is worth keeping since I've found it to be a "canary" that lets me know that something more fundamental is going wrong with my local env (e.g. if I forget `npm run build`).

I tried it merged with `setup`, but then it runs on every test run, which is a bother since `basic` takes a second or so to run while `setup` is instant after the initial auth.

reykjalin reviewed (approved):
LGTM, awesome work 👏 🚀

Jinksi commented:
@achyuthajoy would you still like to re-review this PR?

achyuthajoy committed:
Merge branch 'develop' into try/playwright-e2e-tests

achyuthajoy reviewed (approved):
Thanks for working @Jinksi. Tested the changes and works well.

LGTM 🚢

Jinksi commented:
Thanks! I'll aim to merge this ASAP (~Wednesday~ Friday 🤷).

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Merge branch 'develop' into try/playwright-e2e-tests

Jinksi committed:
Improve wording of useShopper/getShopper difference in README

Jinksi committed:
Remove deprecated `version` from PW `docker-compose.yml`

Jinksi committed:
Upgrade Playwright to latest stable (v1.42.0 → v1.43.1)

Jinksi commented:
Failing `Compatibility` and `Code coverage` GH checks are unrelated – due to #8662.
